core/domain/rights_manager.py

```diff
@@ -826,6 +947,10 @@ def _assign_role(
             rights_domain.ROLE_VIEWER
     ]:
         raise Exception('Invalid role: %s' % new_role)
+
+    # Ruling out the possibility of None for mypy type checking.
```

Comment:
Why is this guaranteed? Also shouldn't the assertion come a lot earlier in this function (noting the change you made to the if clause above)?

Reply:
In line 958, we are accessing the `assign_new_role` method without checking for none case.
So, that's why i added assertion.

Reply:
This reasoning is not valid. See above.

Basically, the code might be wrong and there might actually be a need to handle the None case. An analysis of the callers is needed to figure out whether the code should be changed or whether the assertion is guaranteed to always hold. Bear in mind that new callers may call these functions erroneously (i.e. without preserving any implicit guarantees) so there needs to be safety checks in place.

In general I think that if a variable is not guaranteed to satisfy a particular condition _based on the logic within the function itself_ then you shouldn't assert it, and instead update the code to handle the additional cases.

Reply:
In Line `948`, we are already checking if `activity_rights` is None or not.

In `check_can_modify_core_activity_roles()` method, If we provide `activity_rights` as None then it going to return a False value and that False value is assigned to `user_can_assign_role`. So, if `user_can_assign_role` is False then an error is raised in line `951`.

Reply:
This is reasonable given what check_can_modify_core_activity_roles() currently does. But what if someone modifies that function in the future (perhaps to fix some bug) and doesn't realize that this code is relying on it too?

So that's why I wonder whether you could just add an "if activity_rights is None: raise Exception" above the if clause on line 936, and then assume it's not None from thereon.

Reply:
If clause is added and docstring is updated.